File: web/src/storage/envelope.ts
// Seed envelope: password -> Argon2id (wasm) -> wrap key; a random content
// key encrypts the phrase with AES-256-GCM (WebCrypto). Changing the password
// re-wraps the content key without touching the seed ciphertext (section 6 of
// the architecture document).

import type { SeedEnvelope } from './db';

/** Argon2id parameters; stored in the envelope so they can change later. */
export interface KdfParams {
  mKib: number;
  tCost: number;
  pCost: number;
}

export const DEFAULT_KDF: KdfParams = { mKib: 64 * 1024, tCost: 3, pCost: 1 };

/**
 * The most work an envelope may ask of this device. An envelope arrives
 * from a file as well as from the database, and its numbers go straight to
 * Argon2: without a ceiling, a file asking for gigabytes or billions of
 * passes would hold the wallet worker until the tab was closed. The floor
 * is not here but in the wasm core's Argon2 itself (kdf.rs), so that it
 * binds every real derivation while test doubles stay free to be cheap.
 */
export const KDF_CEILING: KdfParams = { mKib: 1024 * 1024, tCost: 16, pCost: 4 };

/**
 * A backup file larger than this is not one. The size is the contacts':
 * a Standard address is about 3,500 characters, so five thousand contacts,
 * the most a restore takes, come to some 24 MB once encrypted and encoded.
 */
export const MAX_BACKUP_BYTES = 32 * 1024 * 1024;

/** The file says it was changed, or it is damaged: the authenticated part does not verify under the right password. */
export class BackupAlteredError extends Error {
  constructor() {
    super('This backup file has been changed since it was made, or is damaged. Nothing was restored from it.');
    this.name = 'BackupAlteredError';
  }
}

function bytesOf(text: unknown, what: string, maxChars = 4096): Uint8Array {
  if (typeof text !== 'string' || text.length > maxChars) throw new Error(`This wallet data is malformed (${what}).`);
  try {
    return fromB64(text);
  } catch {
    throw new Error(`This wallet data is malformed (${what}).`);
  }
}

function boxLengths(box: unknown, what: string, maxChars = 4096): { iv: number; ciphertext: number } {
  const b = box as { iv?: unknown; ciphertext?: unknown } | null;
  if (typeof b !== 'object' || b === null) throw new Error(`This wallet data is malformed (${what}).`);
  return { iv: bytesOf(b.iv, what).length, ciphertext: bytesOf(b.ciphertext, what, maxChars).length };
}

/**
 * Everything about an envelope that can be checked without the password,
 * checked before any of it is used. The exact lengths matter beyond
 * tidiness: AES-GCM does not commit to one key, and a wrapped key of
 * another length is where a ciphertext built to open under many passwords
 * would hide.
 */
export function assertEnvelope(envelope: unknown): asserts envelope is SeedEnvelope {
  const e = envelope as Partial<SeedEnvelope> | null;
  if (typeof e !== 'object' || e === null) throw new Error('This wallet data is malformed (no envelope).');
  if (e.version !== 1 || e.kdf?.name !== 'argon2id') throw new Error(`unsupported envelope version ${String(e.version)}`);
  const { mKib, tCost, pCost } = e.kdf;
  for (const [n, max] of [[mKib, KDF_CEILING.mKib], [tCost, KDF_CEILING.tCost], [pCost, KDF_CEILING.pCost]] as const) {
    if (!Number.isSafeInteger(n) || n < 1 || n > max) throw new Error('This wallet data asks for a password hash this app will not run: its parameters are out of range.');
  }
  const salt = bytesOf(e.kdf.salt, 'salt').length;
  if (salt < 16 || salt > 64) throw new Error('This wallet data is malformed (salt).');
  const wrapped = boxLengths(e.wrappedContentKey, 'wrapped key');
  // A 32-byte key and AES-GCM's 16-byte tag.
  if (wrapped.iv !== 12 || wrapped.ciphertext !== 48) throw new Error('This wallet data is malformed (wrapped key).');
  const seed = boxLengths(e.seed, 'seed');
  if (seed.iv !== 12 || seed.ciphertext < 17 || seed.ciphertext > 1024) throw new Error('This wallet data is malformed (seed).');
}

/** Whether an envelope's password hash is cheaper to guess against than today's default. */
export function isWeakerThanDefault(kdf: KdfParams, against: KdfParams = DEFAULT_KDF): boolean {
  return kdf.mKib < against.mKib || kdf.tCost < against.tCost;
}

/** Provided by the wasm wallet core (possibly in a worker): Argon2id, 32 bytes. */
export type DeriveKey = (
  password: Uint8Array,
  salt: Uint8Array,
  mKib: number,
  tCost: number,
  pCost: number,
) => Uint8Array | Promise<Uint8Array>;

// WebCrypto wants a plain ArrayBuffer view; copy out of any shared buffer.
function ab(u8: Uint8Array): ArrayBuffer {
  return u8.buffer.slice(u8.byteOffset, u8.byteOffset + u8.byteLength) as ArrayBuffer;
}

const te = new TextEncoder();
const td = new TextDecoder();

function toB64(bytes: Uint8Array): string {
  let s = '';
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}

function fromB64(text: string): Uint8Array {
  const s = atob(text);
  const out = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}

/** WebCrypto exists only in secure contexts (https or localhost). */
function subtle(): SubtleCrypto {
  if (!globalThis.crypto?.subtle) {
    throw new Error('This page is not a secure context, so the browser provides no encryption. Open the app over https or on localhost.');
  }
  return globalThis.crypto.subtle;
}

async function importAesKey(raw: Uint8Array, usages: KeyUsage[]): Promise<CryptoKey> {
  return subtle().importKey('raw', ab(raw), { name: 'AES-GCM' }, false, usages);
}

async function aesEncrypt(key: CryptoKey, plaintext: Uint8Array, additionalData?: Uint8Array): Promise<{ iv: string; ciphertext: string }> {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const params = additionalData ? { name: 'AES-GCM', iv: ab(iv), additionalData: ab(additionalData) } : { name: 'AES-GCM', iv: ab(iv) };
  const ciphertext = new Uint8Array(await subtle().encrypt(params, key, ab(plaintext)));
  return { iv: toB64(iv), ciphertext: toB64(ciphertext) };
}

async function aesDecrypt(key: CryptoKey, box: { iv: string; ciphertext: string }, additionalData?: Uint8Array): Promise<Uint8Array> {
  const iv = ab(fromB64(box.iv));
  const params = additionalData ? { name: 'AES-GCM', iv, additionalData: ab(additionalData) } : { name: 'AES-GCM', iv };
  return new Uint8Array(await subtle().decrypt(params, key, ab(fromB64(box.ciphertext))));
}

/** Build a new envelope for `phrase` under `password`. */
export async function sealSeed(
  phrase: string[],
  password: string,
  deriveKey: DeriveKey,
  kdf: KdfParams = DEFAULT_KDF,
): Promise<SeedEnvelope> {
  const { envelope, contentKey } = await sealSeedKeepingKey(phrase, password, deriveKey, kdf);
  contentKey.fill(0);
  return envelope;
}

/**
 * The same, and the content key with it. The wallet's sealed log is keyed
 * from the content key, so whoever is about to unlock the new wallet needs
 * it once. Zero it when done.
 */
export async function sealSeedKeepingKey(
  phrase: string[],
  password: string,
  deriveKey: DeriveKey,
  kdf: KdfParams = DEFAULT_KDF,
): Promise<{ envelope: SeedEnvelope; contentKey: Uint8Array }> {
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const wrapRaw = await deriveKey(te.encode(password), salt, kdf.mKib, kdf.tCost, kdf.pCost);
  const contentRaw = crypto.getRandomValues(new Uint8Array(32));
  const wrapKey = await importAesKey(wrapRaw, ['encrypt']);
  const contentKey = await importAesKey(contentRaw, ['encrypt']);
  const wrappedContentKey = await aesEncrypt(wrapKey, contentRaw);
  const seed = await aesEncrypt(contentKey, te.encode(phrase.join(' ')));
  wrapRaw.fill(0);
  return {
    envelope: {
      version: 1,
      kdf: { name: 'argon2id', mKib: kdf.mKib, tCost: kdf.tCost, pCost: kdf.pCost, salt: toB64(salt) },
      wrappedContentKey,
      seed,
    },
    contentKey: contentRaw,
  };
}

export class WrongPasswordError extends Error {
  constructor() {
    super('wrong password');
    this.name = 'WrongPasswordError';
  }
}

/** Recover the phrase. Throws WrongPasswordError when the password is wrong. */
export async function openSeed(envelope: SeedEnvelope, password: string, deriveKey: DeriveKey): Promise<string[]> {
  const { phrase, contentKey } = await openSeedKeepingKey(envelope, password, deriveKey);
  contentKey.fill(0);
  return phrase;
}

/** The same, and the content key with it, from the one password hash. Zero it when done. */
export async function openSeedKeepingKey(envelope: SeedEnvelope, password: string, deriveKey: DeriveKey): Promise<{ phrase: string[]; contentKey: Uint8Array }> {
  assertEnvelope(envelope);
  const { mKib, tCost, pCost, salt } = envelope.kdf;
  const wrapRaw = await deriveKey(te.encode(password), fromB64(salt), mKib, tCost, pCost);
  const wrapKey = await importAesKey(wrapRaw, ['decrypt']);
  wrapRaw.fill(0);
  let contentRaw: Uint8Array;
  try {
    contentRaw = await aesDecrypt(wrapKey, envelope.wrappedContentKey);
  } catch {
    throw new WrongPasswordError();
  }
  const contentKey = await importAesKey(contentRaw, ['decrypt']);
  let phrase: Uint8Array;
  try {
    phrase = await aesDecrypt(contentKey, envelope.seed);
  } catch {
    contentRaw.fill(0);
    // The password was right and the seed still does not open: the data is
    // damaged, or it is a newer backup file dressed as an older one.
    throw new Error('The password is right, but the seed in this wallet data does not open: the data is damaged or has been changed.');
  }
  return { phrase: td.decode(phrase).split(' '), contentKey: contentRaw };
}

/** The raw content key, after checking the password. Zero it when done. */
export async function extractContentKey(envelope: SeedEnvelope, password: string, deriveKey: DeriveKey): Promise<Uint8Array> {
  assertEnvelope(envelope);
  const { mKib, tCost, pCost, salt } = envelope.kdf;
  const wrapRaw = await deriveKey(te.encode(password), fromB64(salt), mKib, tCost, pCost);
  const wrapKey = await importAesKey(wrapRaw, ['decrypt']);
  wrapRaw.fill(0);
  try {
    return await aesDecrypt(wrapKey, envelope.wrappedContentKey);
  } catch {
    throw new WrongPasswordError();
  }
}

/** Wrap the content key under a 32-byte secret (a passkey's PRF output). */
export async function wrapContentKey(contentRaw: Uint8Array, secret: Uint8Array): Promise<{ iv: string; ciphertext: string }> {
  const key = await importAesKey(secret, ['encrypt']);
  return aesEncrypt(key, contentRaw);
}

/** Recover the phrase from a content key wrapped under a secret. */
export async function openSeedWithSecret(
  envelope: SeedEnvelope,
  wrapped: { iv: string; ciphertext: string },
  secret: Uint8Array,
): Promise<string[]> {
  const { phrase, contentKey } = await openSeedWithSecretKeepingKey(envelope, wrapped, secret);
  contentKey.fill(0);
  return phrase;
}

/** The same, and the content key with it. Zero it when done. */
export async function openSeedWithSecretKeepingKey(
  envelope: SeedEnvelope,
  wrapped: { iv: string; ciphertext: string },
  secret: Uint8Array,
): Promise<{ phrase: string[]; contentKey: Uint8Array }> {
  assertEnvelope(envelope);
  const key = await importAesKey(secret, ['decrypt']);
  let contentRaw: Uint8Array;
  try {
    contentRaw = await aesDecrypt(key, wrapped);
  } catch {
    throw new Error('This passkey no longer matches the wallet; unlock with the password and set the passkey up again.');
  }
  const contentKey = await importAesKey(contentRaw, ['decrypt']);
  try {
    return { phrase: td.decode(await aesDecrypt(contentKey, envelope.seed)).split(' '), contentKey: contentRaw };
  } catch (e) {
    contentRaw.fill(0);
    throw e;
  }
}

/** Re-wrap the content key under a new password; the seed ciphertext stays. */
export async function changePassword(
  envelope: SeedEnvelope,
  oldPassword: string,
  newPassword: string,
  deriveKey: DeriveKey,
  kdf: KdfParams = DEFAULT_KDF,
): Promise<SeedEnvelope> {
  assertEnvelope(envelope);
  const oldRaw = await deriveKey(te.encode(oldPassword), fromB64(envelope.kdf.salt), envelope.kdf.mKib, envelope.kdf.tCost, envelope.kdf.pCost);
  const oldKey = await importAesKey(oldRaw, ['decrypt']);
  oldRaw.fill(0);
  let contentRaw: Uint8Array;
  try {
    contentRaw = await aesDecrypt(oldKey, envelope.wrappedContentKey);
  } catch {
    throw new WrongPasswordError();
  }
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const newRaw = await deriveKey(te.encode(newPassword), salt, kdf.mKib, kdf.tCost, kdf.pCost);
  const newKey = await importAesKey(newRaw, ['encrypt']);
  const wrappedContentKey = await aesEncrypt(newKey, contentRaw);
  newRaw.fill(0);
  contentRaw.fill(0);
  return {
    ...envelope,
    kdf: { name: 'argon2id', mKib: kdf.mKib, tCost: kdf.tCost, pCost: kdf.pCost, salt: toB64(salt) },
    wrappedContentKey,
  };
}

/**
 * The export file (R8): the seed, protected by the password, plus what is
 * needed to rescan, plus the contacts.
 *
 * Versions 1 and 2 carried the database's envelope as it was and kept
 * everything beside it in clear, bound to nothing. Whoever could write to
 * where the file was kept could swap a contact's address for their own, or
 * raise the start block so that a restore came up empty, without the
 * password and without a trace.
 *
 * Version 3 closes that, in a way an attacker cannot undo by relabelling
 * the file as an older one:
 *
 *   password --Argon2--> wrap key --AES-GCM--> file key
 *   file key --AES-GCM, bound to the readable part--> content key
 *   content key --AES-GCM--> seed            (the same ciphertext as in the database)
 *   content key --AES-GCM, bound to all of the above--> contacts
 *
 * The readable part (format, version, network, start block, export date,
 * the password hash settings) goes into the second step as additional
 * authenticated data: change one byte of it and the content key does not
 * come out, so neither the seed nor the contacts open, and the restore
 * stops. The first step has no such binding, on purpose: it answers one
 * question only, whether the password is right, so that a wrong password
 * and a changed file are told apart.
 *
 * An older reader, handed a version 3 file dressed as version 2, would take
 * the file key for the content key, and the seed would not open under it.
 */
export type ExportFile = LegacyExportFile | SealedExportFile;

/** Versions 1 and 2. */
export interface LegacyExportFile {
  format: 'neptune-vault-backup';
  /** 1: seed only; 2: adds contacts, in clear. */
  version: 1 | 2;
  network: string;
  birthdayHeight: number;
  envelope: SeedEnvelope;
  exportedAt: number;
  contacts?: { name: string; address: string }[];
}

type Box = { iv: string; ciphertext: string };

/** Version 3. */
export interface SealedExportFile {
  format: 'neptune-vault-backup';
  version: 3;
  network: string;
  birthdayHeight: number;
  exportedAt: number;
  envelope: {
    version: 2;
    kdf: SeedEnvelope['kdf'];
    /** The file key under the password's wrap key. Opens for the right password and for nothing else. */
    wrappedFileKey: Box;
    /** The content key under the file key, bound to the readable part of the file. */
    boundContentKey: Box;
    /** The seed under the content key: the database's own ciphertext. */
    seed: Box;
  };
  /** BackupSecrets under the content key, bound to everything above. */
  sealed: Box;
}

/** What a version 3 file keeps encrypted besides the seed. */
export interface BackupSecrets {
  contacts: { name: string; address: string }[];
}

/**
 * The readable part of a version 3 file as one fixed text, byte for byte
 * the same at export and at restore. Built by hand, field by field in a set
 * order: JSON promises neither key order nor number formatting. No field
 * can hold a line break (numbers, fixed names, base64), so the join is
 * unambiguous.
 */
function readablePart(file: Pick<SealedExportFile, 'format' | 'version' | 'network' | 'birthdayHeight' | 'exportedAt'>, kdf: SeedEnvelope['kdf'], wrappedFileKey: Box, seed: Box): string {
  return [
    file.format,
    String(file.version),
    file.network,
    String(file.birthdayHeight),
    String(file.exportedAt),
    '2',
    kdf.name,
    String(kdf.mKib),
    String(kdf.tCost),
    String(kdf.pCost),
    kdf.salt,
    wrappedFileKey.iv,
    wrappedFileKey.ciphertext,
    seed.iv,
    seed.ciphertext,
  ].join('\n');
}

/**
 * Make a version 3 file from the database's envelope. Takes the password:
 * the content key is needed, and being unlocked does not give it. Throws
 * WrongPasswordError. One password hash: the wrap key the password gives
 * for the stored envelope wraps the file key too, under a fresh IV.
 */
export async function sealBackup(
  meta: { network: string; birthdayHeight: number; exportedAt: number },
  envelope: SeedEnvelope,
  secrets: BackupSecrets,
  password: string,
  deriveKey: DeriveKey,
): Promise<SealedExportFile> {
  assertEnvelope(envelope);
  const { mKib, tCost, pCost, salt } = envelope.kdf;
  const wrapRaw = await deriveKey(te.encode(password), fromB64(salt), mKib, tCost, pCost);
  const wrapKey = await importAesKey(wrapRaw, ['encrypt', 'decrypt']);
  wrapRaw.fill(0);
  let contentRaw: Uint8Array;
  try {
    contentRaw = await aesDecrypt(wrapKey, envelope.wrappedContentKey);
  } catch {
    throw new WrongPasswordError();
  }
  const fileRaw = crypto.getRandomValues(new Uint8Array(32));
  try {
    const head = { format: 'neptune-vault-backup' as const, version: 3 as const, ...meta };
    const wrappedFileKey = await aesEncrypt(wrapKey, fileRaw);
    const readable = readablePart(head, envelope.kdf, wrappedFileKey, envelope.seed);
    const boundContentKey = await aesEncrypt(await importAesKey(fileRaw, ['encrypt']), contentRaw, te.encode(readable));
    const everything = readable + '\n' + boundContentKey.iv + '\n' + boundContentKey.ciphertext;
    const sealed = await aesEncrypt(await importAesKey(contentRaw, ['encrypt']), te.encode(JSON.stringify(secrets)), te.encode(everything));
    return { ...head, envelope: { version: 2, kdf: envelope.kdf, wrappedFileKey, boundContentKey, seed: envelope.seed }, sealed };
  } finally {
    contentRaw.fill(0);
    fileRaw.fill(0);
  }
}

/**
 * Open a version 3 file: its secrets, and an ordinary envelope for the
 * database, wrapped under the same password. Throws WrongPasswordError for a
 * wrong password, and BackupAlteredError when the password is right and the
 * file is not what was exported.
 */
export async function openBackup(file: SealedExportFile, password: string, deriveKey: DeriveKey): Promise<{ secrets: BackupSecrets; envelope: SeedEnvelope }> {
  assertFileEnvelope(file);
  const e = file.envelope;
  const wrapRaw = await deriveKey(te.encode(password), fromB64(e.kdf.salt), e.kdf.mKib, e.kdf.tCost, e.kdf.pCost);
  const wrapKey = await importAesKey(wrapRaw, ['encrypt', 'decrypt']);
  wrapRaw.fill(0);
  let fileRaw: Uint8Array;
  try {
    fileRaw = await aesDecrypt(wrapKey, e.wrappedFileKey);
  } catch {
    throw new WrongPasswordError();
  }
  // From here the password is known to be right: whatever fails is the file.
  let contentRaw: Uint8Array | null = null;
  try {
    const readable = readablePart(file, e.kdf, e.wrappedFileKey, e.seed);
    contentRaw = await aesDecrypt(await importAesKey(fileRaw, ['decrypt']), e.boundContentKey, te.encode(readable));
    const everything = readable + '\n' + e.boundContentKey.iv + '\n' + e.boundContentKey.ciphertext;
    const contentKey = await importAesKey(contentRaw, ['decrypt']);
    const plain = await aesDecrypt(contentKey, file.sealed, te.encode(everything));
    // The seed must open under this content key too, or the file would
    // restore contacts to a wallet that then cannot be unlocked.
    await aesDecrypt(contentKey, e.seed);
    const parsed = JSON.parse(td.decode(plain)) as Partial<BackupSecrets>;
    const wrappedContentKey = await aesEncrypt(wrapKey, contentRaw);
    return {
      secrets: { contacts: Array.isArray(parsed.contacts) ? parsed.contacts : [] },
      envelope: { version: 1, kdf: e.kdf, wrappedContentKey, seed: e.seed },
    };
  } catch {
    throw new BackupAlteredError();
  } finally {
    fileRaw.fill(0);
    contentRaw?.fill(0);
  }
}

/** What can be checked of a version 3 file without the password. */
function assertFileEnvelope(file: SealedExportFile): void {
  const e = file.envelope as Partial<SealedExportFile['envelope']> | null;
  if (typeof e !== 'object' || e === null || e.version !== 2 || !e.kdf) throw new BackupAlteredError();
  // The same checks as an ordinary envelope: the settings, the salt, the seed, and a 48-byte wrapped key.
  assertEnvelope({ version: 1, kdf: e.kdf, wrappedContentKey: e.wrappedFileKey, seed: e.seed });
  const bound = boxLengths(e.boundContentKey, 'content key');
  if (bound.iv !== 12 || bound.ciphertext !== 48) throw new Error('This wallet data is malformed (content key).');
  // A version 3 file without its sealed part is one that had it removed.
  if (typeof file.sealed !== 'object' || file.sealed === null) throw new BackupAlteredError();
  // The contacts are in here, and one Standard address alone is several
  // thousand characters: the whole file's cap is the cap.
  const sealed = boxLengths(file.sealed, 'sealed part', MAX_BACKUP_BYTES);
  if (sealed.iv !== 12 || sealed.ciphertext < 17) throw new BackupAlteredError();
  if (typeof file.network !== 'string' || !Number.isSafeInteger(file.birthdayHeight) || !Number.isSafeInteger(file.exportedAt)) throw new BackupAlteredError();
}

/**
 * Read a backup file's text. Everything that can be checked without the
 * password is checked here, before the person is asked for it.
 */
export function parseBackupFile(text: string): ExportFile {
  const notOne = 'This file is not a Neptune Vault backup file.';
  if (text.length > MAX_BACKUP_BYTES) throw new Error('This file is too large to be a Neptune Vault backup file.');
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new Error(notOne);
  }
  const file = parsed as Partial<ExportFile> | null;
  if (typeof file !== 'object' || file === null || file.format !== 'neptune-vault-backup') throw new Error(notOne);
  // Every version ever written stays readable; one this app does not know
  // is from a newer app, never a reason to guess at the contents.
  if (file.version !== 1 && file.version !== 2 && file.version !== 3) {
    throw new Error('This backup file was made by a newer version of Neptune Vault. Update the app, then restore it.');
  }
  if (file.version === 3) assertFileEnvelope(file as SealedExportFile);
  else assertEnvelope(file.envelope);
  return file as ExportFile;
}

File: crates/vault-bridge/src/envelope.rs
//! The seed envelope, opened where the seed can stay.
//!
//! The web app's `storage/envelope.ts` is the reference: it made every
//! envelope that exists, and this must open all of them. The format is two
//! layers of AES-256-GCM over one Argon2id derivation:
//!
//! ```text
//! password --argon2id(salt, params)--> wrap key
//! wrap key --aes-256-gcm--> content key   (the wrappedContentKey box)
//! content key --aes-256-gcm--> the phrase, as UTF-8 words joined by spaces
//! ```
//!
//! The middle layer is what lets a passkey open the same envelope: the
//! content key is wrapped a second time under the passkey's secret, and
//! that box is stored beside the envelope rather than in it.
//!
//! Only the AES layering is written twice, once here and once in
//! TypeScript. Argon2id is not: both sides call `vault_core::kdf`, one
//! through wasm and one directly, so the expensive half of the format
//! cannot drift. The layering is held to the other implementation by
//! `test-vectors/seed-envelope.json`, which both test suites read.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine as _;
use serde::{Deserialize, Serialize};
use zeroize::Zeroizing;

use crate::BridgeError;

type Result<T> = std::result::Result<T, BridgeError>;

/// The most work an envelope may ask of this device.
///
/// An envelope arrives as data, and its parameters say how much memory and
/// time to spend. Without a ceiling, a file could ask for a terabyte of
/// Argon2 memory and take the app down with it. Must match `KDF_CEILING` in
/// `storage/envelope.ts`.
const CEILING_M_KIB: u32 = 1024 * 1024;
const CEILING_T_COST: u32 = 16;
const CEILING_P_COST: u32 = 4;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KdfParams {
    pub name: String,
    #[serde(rename = "mKib")]
    pub m_kib: u32,
    #[serde(rename = "tCost")]
    pub t_cost: u32,
    #[serde(rename = "pCost")]
    pub p_cost: u32,
    /// base64.
    pub salt: String,
}

/// One AES-256-GCM box, both fields base64.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SealedBox {
    pub iv: String,
    pub ciphertext: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SeedEnvelope {
    pub version: u32,
    pub kdf: KdfParams,
    pub wrapped_content_key: SealedBox,
    pub seed: SealedBox,
}

fn malformed(what: &str) -> BridgeError {
    BridgeError::plain(format!("This wallet data is malformed ({what})."))
}

fn wrong_password() -> BridgeError {
    BridgeError::new("WrongPasswordError", "wrong password")
}

fn decode(text: &str, what: &str) -> Result<Vec<u8>> {
    if text.len() > 4096 {
        return Err(malformed(what));
    }
    BASE64.decode(text).map_err(|_| malformed(what))
}

/// Everything about an envelope that can be checked without the password.
///
/// The exact lengths matter beyond tidiness: AES-GCM does not commit to one
/// key, so a wrapped key of another length is where a ciphertext built to
/// open under many passwords would hide.
pub fn assert_envelope(envelope: &SeedEnvelope) -> Result<()> {
    if envelope.version != 1 || envelope.kdf.name != "argon2id" {
        return Err(BridgeError::plain(format!(
            "unsupported envelope version {}",
            envelope.version
        )));
    }
    let kdf = &envelope.kdf;
    let within = kdf.m_kib >= 1
        && kdf.m_kib <= CEILING_M_KIB
        && kdf.t_cost >= 1
        && kdf.t_cost <= CEILING_T_COST
        && kdf.p_cost >= 1
        && kdf.p_cost <= CEILING_P_COST;
    if !within {
        return Err(BridgeError::plain(
            "This wallet data asks for a password hash this app will not run: its parameters are out of range.",
        ));
    }

    let salt = decode(&kdf.salt, "salt")?.len();
    if !(16..=64).contains(&salt) {
        return Err(malformed("salt"));
    }

    // A 32-byte key and AES-GCM's 16-byte tag.
    let iv = decode(&envelope.wrapped_content_key.iv, "wrapped key")?.len();
    let ciphertext = decode(&envelope.wrapped_content_key.ciphertext, "wrapped key")?.len();
    if iv != 12 || ciphertext != 48 {
        return Err(malformed("wrapped key"));
    }

    let iv = decode(&envelope.seed.iv, "seed")?.len();
    let ciphertext = decode(&envelope.seed.ciphertext, "seed")?.len();
    if iv != 12 || !(17..=1024).contains(&ciphertext) {
        return Err(malformed("seed"));
    }
    Ok(())
}

/// Open one box under a 32-byte key.
fn open_box(key: &[u8], sealed: &SealedBox, what: &str) -> Result<Zeroizing<Vec<u8>>> {
    let key: [u8; 32] = key.try_into().map_err(|_| malformed(what))?;
    let cipher = Aes256Gcm::new(&key.into());
    let iv = decode(&sealed.iv, what)?;
    let ciphertext = decode(&sealed.ciphertext, what)?;
    let plaintext = cipher
        .decrypt(
            Nonce::from_slice(&iv),
            Payload {
                msg: &ciphertext,
                aad: &[],
            },
        )
        .map_err(|_| malformed(what))?;
    Ok(Zeroizing::new(plaintext))
}

/// The wrap key for an envelope, from a password. The expensive step.
fn wrap_key(envelope: &SeedEnvelope, password: &str) -> Result<Zeroizing<[u8; 32]>> {
    let salt = decode(&envelope.kdf.salt, "salt")?;
    vault_core::kdf::derive_key(
        password.as_bytes(),
        &salt,
        envelope.kdf.m_kib,
        envelope.kdf.t_cost,
        envelope.kdf.p_cost,
    )
    .map_err(Into::into)
}

/// The content key, after checking the password.
pub fn content_key(envelope: &SeedEnvelope, password: &str) -> Result<Zeroizing<Vec<u8>>> {
    assert_envelope(envelope)?;
    let wrap = wrap_key(envelope, password)?;
    // The only thing a wrong password can look like: the wrapped key does
    // not authenticate. Everything past here is damage, not a typo.
    open_box(wrap.as_slice(), &envelope.wrapped_content_key, "wrapped key")
        .map_err(|_| wrong_password())
}

/// The content key, from a passkey's secret instead of a password.
///
/// `wrapped` is the same content key sealed under the secret, which the app
/// stores beside the envelope when a passkey is enrolled.
pub fn content_key_from_secret(
    wrapped: &SealedBox,
    secret: &[u8],
) -> Result<Zeroizing<Vec<u8>>> {
    open_box(secret, wrapped, "passkey key").map_err(|_| wrong_password())
}

/// The seed phrase, given the content key.
pub fn phrase_from_content_key(
    envelope: &SeedEnvelope,
    content_key: &[u8],
) -> Result<Vec<String>> {
    let plaintext = open_box(content_key, &envelope.seed, "seed").map_err(|_| {
        // The password was right and the seed still does not open: the data
        // is damaged, or it is a newer backup file dressed as an older one.
        BridgeError::plain(
            "The password is right, but the seed in this wallet data does not open: the data is damaged or has been changed.",
        )
    })?;
    let text = std::str::from_utf8(&plaintext).map_err(|_| malformed("seed"))?;
    Ok(text.split(' ').map(str::to_string).collect())
}

/// Recover the phrase from an envelope and its password.
pub fn open_seed(envelope: &SeedEnvelope, password: &str) -> Result<Vec<String>> {
    let content = content_key(envelope, password)?;
    phrase_from_content_key(envelope, &content)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The vector both implementations are held to. Only the AES layering
    /// is duplicated, so the vector fixes the wrap key rather than a
    /// password: what it pins is how the two boxes nest, which is exactly
    /// what could drift.
    #[derive(Deserialize)]
    struct Vector {
        /// base64, what Argon2id would have produced.
        wrap_key: String,
        envelope: SeedEnvelope,
        phrase: Vec<String>,
        /// The content key wrapped under a passkey secret, and that secret.
        passkey_wrapped: SealedBox,
        passkey_secret: String,
    }

    fn vector() -> Vector {
        let path = concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../test-vectors/seed-envelope.json"
        );
        let text = std::fs::read_to_string(path).expect("the shared test vector");
        serde_json::from_str(&text).expect("the shared test vector parses")
    }

    #[test]
    fn opens_the_shared_vector() {
        let v = vector();
        let wrap = BASE64.decode(&v.wrap_key).unwrap();
        assert_envelope(&v.envelope).unwrap();
        let content = open_box(&wrap, &v.envelope.wrapped_content_key, "wrapped key").unwrap();
        assert_eq!(phrase_from_content_key(&v.envelope, &content).unwrap(), v.phrase);
    }

    #[test]
    fn opens_the_shared_vector_through_a_passkey() {
        let v = vector();
        let secret = BASE64.decode(&v.passkey_secret).unwrap();
        let content = content_key_from_secret(&v.passkey_wrapped, &secret).unwrap();
        assert_eq!(phrase_from_content_key(&v.envelope, &content).unwrap(), v.phrase);
    }

    #[test]
    fn a_wrong_wrap_key_is_a_wrong_password() {
        let v = vector();
        let e = content_key(&v.envelope, "not the password").unwrap_err();
        assert_eq!(e.name, "WrongPasswordError");
    }

    #[test]
    fn work_beyond_the_ceiling_is_refused() {
        let mut v = vector();
        v.envelope.kdf.m_kib = CEILING_M_KIB + 1;
        let e = assert_envelope(&v.envelope).unwrap_err();
        assert!(e.message.contains("out of range"), "{}", e.message);
    }

    #[test]
    fn a_wrapped_key_of_another_length_is_refused() {
        let mut v = vector();
        v.envelope.wrapped_content_key.ciphertext = BASE64.encode([0u8; 64]);
        assert!(assert_envelope(&v.envelope).is_err());
    }

    #[test]
    fn a_later_version_is_refused() {
        let mut v = vector();
        v.envelope.version = 2;
        let e = assert_envelope(&v.envelope).unwrap_err();
        assert!(e.message.contains("version 2"), "{}", e.message);
    }
}
